File: apps/dashboard/src/types/incident.ts
export interface IncidentListItem {
  id: string
  exception_type: string
  exception_message: string
  priority: string
  status: string
  created_at: string
  updated_at: string
  has_analysis: boolean
  pr_url: string | null
}

export interface AgentTraceEntry {
  agent_name: string
  prompt_version: string
  input_summary: string
  output_summary: string
  latency_ms: number
  error: string | null
}

export interface Recommendation {
  rank: number
  title: string
  description: string
  affected_files: string[]
  suggested_change: string
  confidence: number
  source_refs: string[]
}

export interface RootCauseJson {
  component: string
  likely_cause: string
  contributing_factors: string[]
  confidence: number
}

export interface IncidentDetail {
  id: string
  exception_type: string
  exception_message: string
  stack_trace: string | null
  priority: string
  status: string
  created_at: string
  updated_at: string
  root_cause: string | null
  root_cause_json: RootCauseJson | null
  recommendations: Recommendation[]
  code_snippets: unknown[]
  rag_results: unknown[]
  agent_trace: AgentTraceEntry[]
  // Phase 19 — approval gate
  approval_status: string | null
  approved_by: string | null
  approved_at: string | null
  approved_recommendation_rank: number | null
  pr_url: string | null
  pr_branch: string | null
}

export interface ApprovalResponse {
  incident_id: string
  approval_status: string
  approved_recommendation_rank: number | null
  approved_by: string | null
  approved_at: string | null
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
  pages: number
}

File: apps/api/schemas/incident.py
from __future__ import annotations

from datetime import datetime
from math import ceil
from typing import Any, TypeVar
from uuid import UUID

from pydantic import BaseModel

T = TypeVar("T")


class PaginatedResponse[T](BaseModel):
    items: list[T]
    total: int
    page: int
    page_size: int
    pages: int

    @classmethod
    def build(
        cls,
        items: list[T],
        total: int,
        page: int,
        page_size: int,
    ) -> PaginatedResponse[T]:
        return cls(
            items=items,
            total=total,
            page=page,
            page_size=page_size,
            pages=ceil(total / page_size) if total > 0 else 0,
        )


class IncidentListItem(BaseModel):
    id: UUID
    exception_type: str
    exception_message: str
    priority: str
    status: str
    created_at: datetime
    updated_at: datetime
    has_analysis: bool
    pr_url: str | None = None


class IncidentDetail(BaseModel):
    id: UUID
    exception_type: str
    exception_message: str
    stack_trace: str | None = None
    priority: str
    status: str
    created_at: datetime
    updated_at: datetime
    root_cause: str | None = None
    root_cause_json: dict[str, Any] | None = None
    recommendations: list[dict[str, Any]] = []
    code_snippets: list[dict[str, Any]] = []
    rag_results: list[dict[str, Any]] = []
    agent_trace: list[dict[str, Any]] = []
    approval_status: str | None = None
    approved_by: str | None = None
    approved_at: datetime | None = None
    approved_recommendation_rank: int | None = None
    pr_url: str | None = None
    pr_branch: str | None = None
